crates/shared/src/price_estimation/competition.rs

```diff
@@ -0,0 +1,148 @@
+use super::{Estimate, PriceEstimating, PriceEstimationError, Query};
+use anyhow::{anyhow, Result};
+use futures::future;
+
+/// Price estimator that pulls estimates from various sources
+/// and competes on the best price.
+pub struct CompetitionPriceEstimator {
+    inner: Vec<(String, Box<dyn PriceEstimating>)>,
+}
+
+impl CompetitionPriceEstimator {
+    pub fn new(inner: Vec<(String, Box<dyn PriceEstimating>)>) -> Self {
+        assert!(!inner.is_empty());
+        Self { inner }
+    }
+}
+
+#[async_trait::async_trait]
+impl PriceEstimating for CompetitionPriceEstimator {
+    async fn estimates(&self, queries: &[Query]) -> Vec<Result<Estimate, PriceEstimationError>> {
+        let all_estimates =
+            future::join_all(self.inner.iter().map(|(name, estimator)| async move {
+                (name, estimator.estimates(queries).await)
+            }))
+            .await;
+
+        queries
+            .iter()
+            .enumerate()
+            .map(|(i, query)| {
+                all_estimates
+                    .iter()
+                    .filter_map(|(name, estimates)| match &estimates[i] {
+                        Ok(estimate) => Some((name, estimate)),
+                        Err(err) => {
+                            tracing::warn!(
+                                estimator_name = %name, ?query, ?err,
+                                "price estimation error",
+                            );
+                            None
+                        }
+                    })
+                    .filter_map(|(name, estimate)| {
+                        match estimate.price_in_sell_token_rational(query) {
+                            Some(price) => Some((name, estimate, price)),
+                            None => {
+                                tracing::warn!(
+                                    estimator_name = %name, ?query, ?estimate,
+                                    "price estimate with zero amounts",
+                                );
+                                None
+                            }
+                        }
+                    })
+                    .max_by_key(|(_, _, price)| price.clone())
```

Comment:
Reading this snippet is a great example of why rust's type system is simply awesome. Forces you to think about everything that could go wrong 🤟 

In ts it would be too easy to just have a giant try catch block around all of the logic